warn/scrapers/la.py

```diff
@@ -0,0 +1,394 @@
+
+    Keyword arguments:
+    data_dir -- the Path were the result will be saved (default WARN_DATA_DIR)
+    cache_dir -- the Path where results can be cached (default WARN_CACHE_DIR)
+
+    Returns: the Path where the file is written
+    """
+    # Fire up the cache
+    cache = Cache(cache_dir)
+
+    # The basic configuration for the scrape
+    state_code = "la"
+    base_url = "https://www.laworks.net/"
+    file_base = "Downloads_WFD"
+
+    # Download the root page
+    url = f"{base_url}Downloads/{file_base}.asp"
+    html = utils.get_url(url).text
+
+    # Save it to the cache
+    cache_key = f"{state_code}/{file_base}.html"
+    cache.write(cache_key, html)
+
+    # Parse out the links to WARN notice PDFs
+    links = _parse_links(html)
+
+    all_rows = []
+
+    for link in links:
+        if "WARN Notices" in link.text:
+            # Download the PDF
+            pdf_url = f"{base_url}{link['href']}"
+            pdf_path = _read_or_download(cache, state_code, pdf_url)
+
+            # Process the PDF
+            rows = _process_pdf(pdf_path)
+            all_rows.extend(rows)
+
+    # Insert a header row with clean column names.
+    # We are here assuming that the columns don't change between years
+    # and that one that contains "Employees Affected" will be clean.
+    output_rows = [list(filter(_is_clean_header, all_rows))[0]] + list(
+        filter(_is_not_header, all_rows)
+    )
+
+    # Write out to CSV
+    data_path = data_dir / f"{state_code}.csv"
+    utils.write_rows_to_csv(data_path, output_rows)
+
+    # Return the path
+    return data_path
+
+
+def _process_pdf(pdf_path: Path) -> list:
+    """
+    Process a PDF file.
+
+    Keyword arguments:
+    pdf_path -- the path to the PDF file
+
+    Returns: a list of rows
+    """
+    output_rows: list = []
+
+    with pdfplumber.open(pdf_path) as pdf:
+        for page in pdf.pages:
+            for table in page.debug_tablefinder().tables:
+                for index, row in enumerate(table.rows):
+                    row = [_extract_cell_chars(page, cell) for cell in row.cells]
+
+                    # If the first row in a table appears to be carried over from
+                    # a prior page, (as indicated by mostly blank cells), append the
+                    # row to the previous row
+                    if (
+                        index == 0
+                        and len(list(filter(pdfplumber.utils.extract_text, row))) <= 2
+                        and len(output_rows) > 0
+                    ):
+                        for column_index, cell in enumerate(row):
+                            if column_index < len(output_rows[len(output_rows) - 1]):
+                                output_rows[len(output_rows) - 1][column_index].extend(
+                                    cell
+                                )
+                    # Otherwise, append the row
+                    else:
+                        output_rows.append(row)
```

Comment:
Feels like this could be better factored.

Reply:
As someone who doesn't understand the core issue, it's def. hard to parse. I don't have the perfect solution off the top of my head, but I think so that techniques like breaking out clearly named true/false functions, and breaking it up into a series of if statements rather than one mega clause, tends to help.

Reply:
It's handling this one line:
<img width="1347" alt="Screen Shot 2022-01-28 at 1 07 30 AM" src="https://user-images.githubusercontent.com/1103467/151496038-07cec007-af94-406c-adda-699564b963ed.png">

Reply:
But if you don't understand from looking at the code, I haven't done my job yet.

Reply:
PDFs are the devil!

Reply:
Tried to clean this up.